api/v1alpha1/tinkerbell_type.go

```diff
@@ -0,0 +1,125 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:object:root=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=tinkerbell,scope=Namespaced,categories=tinkerbell,singular=tinkerbell,shortName=tb
+// +kubebuilder:storageversion
+
+// Tinkerbell represents the tinkerbell stack that is being deployed in the kubernetes where the operator is deployed.
+// Tinkerbell operator watches for different resources such as deployment, services, serviceAccounts, etc. One of those
+// CRs is Tinkerbell which the operator will install the tink-stack based on its specs. Once the CR is deleted,
+// the operator will delete all tinkerbell resources.
+type Tinkerbell struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec describes the desired tinkerbell stack state.
+	Spec TinkerbellSpec `json:"spec,omitempty"`
+}
+
+// TinkerbellSpec specifies details of an Tinkerbell setup.
+type TinkerbellSpec struct {
+	// TinkerbellVersion is the tinkerbell crd version.
+	TinkerbellVersion string `json:"tinkerbellVersion"`
+
+	// ClusterDNS is the ip address of the cluster dns resolver.
+	ClusterDNS string `json:"clusterDNS"`
+
+	// OverwriteRegistry is the registry to use for all images. If this field is set, all tink service deployment images
+	// will be prefixed with this value. For example if the value here was set to docker.io, then boots image will be
+	// docker.io/tinkerbell/boots.
+	OverwriteRegistry string `json:"overwriteRegistry"`
+
+	// DockerPullConfig the secret name containing the docker auth config which should exist in the same namespace where
+	// the operator is deployed(typically tinkerbell)
+	DockerPullComfig string `json:"dockerPullComfig"`
+
+	// Boots contains all the information and spec about boots.
+	Boots BootsSpec `json:"boots"`
+	// Hegel contains all the information and spec about boots.
+	Hegel HegelSpec `json:"hegel"`
+	// Rufio contains all the information and spec about rufio.
+	Rufio RufioSpec `json:"rufio"`
+	// Tink contains all the information and spec about tink.
+	Tink TinkSpec `json:"tink"`
+}
+
+// ImageSpec specifies the details of a tinkerbell services images.
+type ImageSpec struct {
```

Comment:
How does this intertwine with the `OverwriteRegistry` configuration in `TinkerbellSpec`? For example, a user could use registry A for Tink, Boots and Hegel but want registry B for Rufio.

Reply:
Well, this won't work. If Registry field was empty, we will use a default registry which is gonna be quay.io and it will be used for services. However, if the registry field was to docker.io then we pull from there instead. Having a field to specify a registry per image in a deployment is kinda over kill. especially that, those are exclusive images to a specific Org. Unlike having to run lots of images from different vendors like workflow actions ;)

Reply:
What is `ImageSpec.ImageRepository`? Is it a genuine Docker repository (some people include the registry with the repository portion)?

For example, we would expect users to configure `chrisdoherty4/hegel` as opposed to `localregistry.io/chrisdoherty4/hegel`? If they do need a custom registry, they'd override that separately with the top level field?

Reply:
`ImageRepository` is just the repo name. For example `quay.io/tinkerbell/tink-worker:v0.9.0`: 
Registry: `quay,io`
ImageRepository: `tinkerbell/tink-worker`
Tag: `v0.9.0`

The full canonical url isn't considered part of the repo. So this struct here:
```go
hegel := Image{
  Repository: "tinkerbell/tink-worker",
  Tag: "v0.9.0"
}
```

By default(without setting the Registry field) Will be:
`quay.io/tinkerbell/tink-worker:v0.9.0"`

if Registry was set to `localregistry.io`, then it will be:
`localregistry.io/tinkerbell/tink-worker:v0.9.0"`